# Machine Learning

Der Prozess, mit einem unterliegenden Datenset einen Algorithmus (das Model) zu erzeugen

In dem Datenset befinden sich Zusammenhänge; der Algorithmus soll diese Zusammenhänge selbstständig erkennen und speichern

Model: Fertiges Programm, welches beim ML Prozess erzeugt wird, und die Aussagen zu neuen Daten treffen kann

Beispiel Wetter-Modell:

Inputs: Datum, Uhrzeit, Wolken%, Niederschlag%

Output: Temperatur

20.12.2024, 14:00, 100%, 0% -> 2°C, 5°C, 7°C, 6°C; Realität: 6°C

Bei 4 Parametern kann dieser Algorithmus noch per Hand geschrieben werden, bei 10+ Parametern ist dies schon schwieriger

---

Wie funktioniert ChatGPT?

Wir geben eine Frage ein, diese Frage enthält Begriffe

Diese Begriffe haben Zusammenhänge mit anderen Begriffen

Die Begriffe werden weiterverfolgt nach weiteren Begriffen

Diese Begriffe werden am Ende zu einem lesbaren Text zusammengebaut

Unter ChatGPT befinden sich 58TB an Daten, das Modell hat 1.2Mrd. Parameter (1 Parameter = 1 Spalte in einer Tabelle)

## Datenset

Das Income.csv Datenset enthält Personendaten, welche aussagen, ob die entsprechende Person unter/über 50.000$/Jahr verdient

Bei diesem Datenset geht es um Binäre Klassifizierung

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv("Data/Income.csv")

In [3]:
data

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race  gender  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country  class  
0              2174             0              40  United-States  <=50K  
1                 0             0              13  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40           Cuba  <=50K  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States  <=50K  
32557             0             0              40  United-States   >50K  
32558             0             0              40  United-States  <=50K  
32559             0             0              20  United-States  <=50K  
32560         15024             0              40  United-States   >50K  

[32561 rows x 15 columns]

## Probleme mit dem Datenset

- Alle Spalten dürfen nur numerische Werte enthalten
- Unebene Daten (capital-gain: 0, 15024, 2174, ...)
- Ungleiche Datenmengen anhand der Klasse

### Daten numerisch machen

In [8]:
data["class"] == ">50K"

0        False
1        False
2        False
3        False
4        False
         ...  
32556    False
32557     True
32558    False
32559    False
32560     True
Name: class, Length: 32561, dtype: bool

In [9]:
(data["class"] == ">50K").astype(int)

0        0
1        0
2        0
3        0
4        0
        ..
32556    0
32557    1
32558    0
32559    0
32560    1
Name: class, Length: 32561, dtype: int32

In [10]:
data["class"] = (data["class"] == ">50K").astype(int)

In [11]:
data

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race  gender  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country  class  
0              2174             0              40  United-States      0  
1                 0             0              13  United-States      0  
2                 0             0              40  United-States      0  
3                 0             0              40  United-States      0  
4                 0             0              40           Cuba      0  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States      0  
32557             0             0              40  United-States      1  
32558             0             0              40  United-States      0  
32559             0             0              20  United-States      0  
32560         15024             0              40  United-States      1  

[32561 rows x 15 columns]

Um alle Spalten numerisch zu machen, können wir den LabelEncoder benutzen

In [12]:
from sklearn.preprocessing import LabelEncoder

In [13]:
enc = LabelEncoder()

In [15]:
enc.fit_transform(data["workclass"])  # Wandelt eine Spalte in eine numerische Spalte um

array([7, 6, 4, ..., 4, 4, 5])

In [18]:
data["workclass"] = enc.fit_transform(data["workclass"])

In [19]:
data

age  workclass  fnlwgt   education  education-num      marital-status  \
0       39          7   77516   Bachelors             13       Never-married   
1       50          6   83311   Bachelors             13  Married-civ-spouse   
2       38          4  215646     HS-grad              9            Divorced   
3       53          4  234721        11th              7  Married-civ-spouse   
4       28          4  338409   Bachelors             13  Married-civ-spouse   
...    ...        ...     ...         ...            ...                 ...   
32556   27          4  257302  Assoc-acdm             12  Married-civ-spouse   
32557   40          4  154374     HS-grad              9  Married-civ-spouse   
32558   58          4  151910     HS-grad              9             Widowed   
32559   22          4  201490     HS-grad              9       Never-married   
32560   52          5  287927     HS-grad              9  Married-civ-spouse   

              occupation   relationship   race  gender  capital-gain  \
0           Adm-clerical  Not-in-family  White    Male          2174   
1        Exec-managerial        Husband  White    Male             0   
2      Handlers-cleaners  Not-in-family  White    Male             0   
3      Handlers-cleaners        Husband  Black    Male             0   
4         Prof-specialty           Wife  Black  Female             0   
...                  ...            ...    ...     ...           ...   
32556       Tech-support           Wife  White  Female             0   
32557  Machine-op-inspct        Husband  White    Male             0   
32558       Adm-clerical      Unmarried  White  Female             0   
32559       Adm-clerical      Own-child  White    Male             0   
32560    Exec-managerial           Wife  White  Female         15024   

       capital-loss  hours-per-week native-country  class  
0                 0              40  United-States      0  
1                 0              13  United-States      0  
2                 0              40  United-States      0  
3                 0              40  United-States      0  
4                 0              40           Cuba      0  
...             ...             ...            ...    ...  
32556             0              38  United-States      0  
32557             0              40  United-States      1  
32558             0              40  United-States      0  
32559             0              20  United-States      0  
32560             0              40  United-States      1  

[32561 rows x 15 columns]

In [20]:
def encodeColumn(columnName: str):
    data[columnName] = enc.fit_transform(data[columnName])

In [31]:
for col in data.select_dtypes("object").columns:  # select_dtypes: Spalten auswählen anhand von Typen
    encodeColumn(col)

In [32]:
data

age  workclass  fnlwgt  education  education-num  marital-status  \
0       39          7   77516          9             13               4   
1       50          6   83311          9             13               2   
2       38          4  215646         11              9               0   
3       53          4  234721          1              7               2   
4       28          4  338409          9             13               2   
...    ...        ...     ...        ...            ...             ...   
32556   27          4  257302          7             12               2   
32557   40          4  154374         11              9               2   
32558   58          4  151910         11              9               6   
32559   22          4  201490         11              9               4   
32560   52          5  287927         11              9               2   

       occupation  relationship  race  gender  capital-gain  capital-loss  \
0               1             1     4       1          2174             0   
1               4             0     4       1             0             0   
2               6             1     4       1             0             0   
3               6             0     2       1             0             0   
4              10             5     2       0             0             0   
...           ...           ...   ...     ...           ...           ...   
32556          13             5     4       0             0             0   
32557           7             0     4       1             0             0   
32558           1             4     4       0             0             0   
32559           1             3     4       1             0             0   
32560           4             5     4       0         15024             0   

       hours-per-week  native-country  class  
0                  40              39      0  
1                  13              39      0  
2                  40              39      0  
3                  40              39      0  
4                  40               5      0  
...               ...             ...    ...  
32556              38              39      0  
32557              40              39      1  
32558              40              39      0  
32559              20              39      0  
32560              40              39      1  

[32561 rows x 15 columns]

### Unebene Daten

Um unebene Daten auszubessen, können wir aus sklearn den StandardScaler verwenden

In [33]:
from sklearn.preprocessing import StandardScaler

In [34]:
scaler = StandardScaler()

In [44]:
toScale = data.iloc[:, :-1]

In [47]:
scaled = scaler.fit_transform(toScale)

In [53]:
data["class"].values.reshape(-1, 1)  # Die Spalte class muss hier 2D werden

array([[0],
       [0],
       [0],
       ...,
       [0],
       [0],
       [1]])

In [52]:
np.hstack((scaled, data["class"].values.reshape(-1, 1)))

array([[ 0.03067056,  2.15057856, -1.06361075, ..., -0.03542945,
         0.29156857,  0.        ],
       [ 0.83710898,  1.46373585, -1.008707  , ..., -2.22215312,
         0.29156857,  0.        ],
       [-0.04264203,  0.09005041,  0.2450785 , ..., -0.03542945,
         0.29156857,  0.        ],
       ...,
       [ 1.42360965,  0.09005041, -0.35877741, ..., -0.03542945,
         0.29156857,  0.        ],
       [-1.21564337,  0.09005041,  0.11095988, ..., -1.65522476,
         0.29156857,  0.        ],
       [ 0.98373415,  0.77689313,  0.92989258, ..., -0.03542945,
         0.29156857,  1.        ]])

In [54]:
data = pd.DataFrame(np.hstack((scaled, data["class"].values.reshape(-1, 1))))

In [55]:
data

0         1         2         3         4         5         6   \
0      0.030671  2.150579 -1.063611 -0.335437  1.134739  0.921634 -1.317809   
1      0.837109  1.463736 -1.008707 -0.335437  1.134739 -0.406212 -0.608387   
2     -0.042642  0.090050  0.245079  0.181332 -0.420060 -1.734058 -0.135438   
3      1.057047  0.090050  0.425801 -2.402511 -1.197459 -0.406212 -0.135438   
4     -0.775768  0.090050  1.408176 -0.335437  1.134739 -0.406212  0.810458   
...         ...       ...       ...       ...       ...       ...       ...   
32556 -0.849080  0.090050  0.639741 -0.852205  0.746039 -0.406212  1.519881   
32557  0.103983  0.090050 -0.335433  0.181332 -0.420060 -0.406212  0.101036   
32558  1.423610  0.090050 -0.358777  0.181332 -0.420060  2.249480 -1.317809   
32559 -1.215643  0.090050  0.110960  0.181332 -0.420060  0.921634 -1.317809   
32560  0.983734  0.776893  0.929893  0.181332 -0.420060 -0.406212 -0.608387   

             7         8         9         10       11        12        13  \
0     -0.277805  0.393668  0.703071  0.148453 -0.21666 -0.035429  0.291569   
1     -0.900181  0.393668  0.703071 -0.145920 -0.21666 -2.222153  0.291569   
2     -0.277805  0.393668  0.703071 -0.145920 -0.21666 -0.035429  0.291569   
3     -0.900181 -1.962621  0.703071 -0.145920 -0.21666 -0.035429  0.291569   
4      2.211698 -1.962621 -1.422331 -0.145920 -0.21666 -0.035429 -4.054223   
...         ...       ...       ...       ...      ...       ...       ...   
32556  2.211698  0.393668 -1.422331 -0.145920 -0.21666 -0.197409  0.291569   
32557 -0.900181  0.393668  0.703071 -0.145920 -0.21666 -0.035429  0.291569   
32558  1.589322  0.393668 -1.422331 -0.145920 -0.21666 -0.035429  0.291569   
32559  0.966947  0.393668  0.703071 -0.145920 -0.21666 -1.655225  0.291569   
32560  2.211698  0.393668 -1.422331  1.888424 -0.21666 -0.035429  0.291569   

        14  
0      0.0  
1      0.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
32556  0.0  
32557  1.0  
32558  0.0  
32559  0.0  
32560  1.0  

[32561 rows x 15 columns]

### Aufteilung der Daten

Für den Trainingsprozess werden Trainingsdaten und nach dem Trainingsprozess Testdaten benötigt

Hierfür wird das originale datenset in 80% Trainingsdaten und 20% Testdaten aufgeteilt

In [59]:
t = int(len(data) * 0.8)

In [61]:
t

26048

In [60]:
len(data) - t

6513

In [63]:
data.sample()  # sample: Ein zufälliger Datensatz

0        1        2         3         4         5         6   \
22479 -1.582206  0.09005 -0.05287 -2.402511 -1.197459  0.921634  1.283407   

             7         8         9        10       11        12        13   14  
22479  0.966947  0.393668 -1.422331 -0.14592 -0.21666 -2.222153  0.291569  0.0

In [66]:
random = data.sample(frac=1)  # Mit frac=1: Datenset mischen

In [67]:
training = random.iloc[0:t]

In [68]:
test = random.iloc[t:]

In [69]:
len(training)

26048

In [70]:
len(test)

6513

### Unebenheiten der Klasse beheben

Wenn im Datenset eine ungleiche Menge an Daten vorhanden ist, kann das Modell eine Neigung (Bias) entwickeln

In [74]:
training[14].value_counts()

14
0.0    19721
1.0     6327
Name: count, dtype: int64

In [75]:
from imblearn.over_sampling import RandomOverSampler

In [76]:
ros = RandomOverSampler()

In [79]:
left, right = ros.fit_resample(training.iloc[:, :-1], training.iloc[:, -1])  # 2 Parameter: Daten ohne class, class alleine

In [85]:
training = pd.DataFrame(np.hstack((left, right.values.reshape(-1, 1))))

In [86]:
training[14].value_counts()

14
0.0    19721
1.0    19721
Name: count, dtype: int64

In [89]:
training = training.sample(frac=1)

#### Aufteilung Links/Rechts

In [90]:
trainingL = training.iloc[:, :-1]
trainingR = training.iloc[:, -1]

testL = test.iloc[:, :-1]
testR = test.iloc[:, -1]

## Simple Algorithmen

### kNN

k-nearest neighbors

- Neuer Datenpunkt ohne Klasse wird platziert
- Jetzt wird geprüft, wieviele Punkte von welcher Klasse näher an dem Punkt sind
- Die Klasse die mehr Punkte in der nähe hat, gibt die Klasse an den neuen Punkt weiter

In [94]:
from sklearn.neighbors import KNeighborsClassifier

In [97]:
knn = KNeighborsClassifier(7)  # k = 7

In [98]:
knn_model = knn.fit(trainingL, trainingR)

In [99]:
knn_model

KNeighborsClassifier(n_neighbors=7)

In [102]:
predictions = knn_model.predict(testL)  # Vorhersagen aus dem Model mittels predict generieren lassen

In [103]:
predictions

array([1., 0., 1., ..., 1., 0., 0.])

In [105]:
testR.values

array([1., 0., 0., ..., 1., 0., 0.])

In [106]:
predictions == testR.values

array([ True,  True, False, ...,  True,  True,  True])

In [108]:
pd.Series(predictions == testR.values).value_counts()

True     5053
False    1460
Name: count, dtype: int64

In [110]:
def evaluate(prediction):
    counts = pd.Series(predictions == testR.values).value_counts()
    print(f"Korrekte Vorhersagen: {counts[True] / len(testR) * 100}%")
    print(f"Falsche Vorhersagen: {counts[False] / len(testR) * 100}%")

In [112]:
evaluate(predictions)

Korrekte Vorhersagen: 77.5832949485644%
Falsche Vorhersagen: 22.41670505143559%


In [116]:
pd.DataFrame(np.hstack((testL, testR.values.reshape(-1, 1), predictions.reshape(-1, 1))))

0         1         2         3         4         5         6   \
0     0.177296  0.090050 -0.804241  0.439716  1.523438 -0.406212 -0.608387   
1    -0.775768  0.090050 -0.491532  0.181332 -0.420060 -1.734058 -1.317809   
2     0.470546  0.090050 -0.019407 -0.335437  1.134739  0.921634  0.810458   
3    -0.775768  0.090050 -0.062790  0.181332 -0.420060 -0.406212 -0.844861   
4     1.570235  2.150579 -0.351188 -0.335437  1.134739 -0.406212 -0.608387   
...        ...       ...       ...       ...       ...       ...       ...   
6508  1.130359  0.090050  1.230892  1.214869 -0.031360 -0.406212  0.810458   
6509  0.250608  0.090050  1.904365 -0.335437  1.134739 -0.406212 -1.317809   
6510  1.643547  1.463736 -0.726666 -1.110590 -1.974858 -0.406212  1.283407   
6511 -0.629143  0.090050 -1.478946  0.181332 -0.420060  0.921634  1.283407   
6512 -0.555830  0.090050  1.069156  1.214869 -0.031360 -1.734058 -1.317809   

            7         8         9        10       11        12        13   14  \
0    -0.900181  0.393668  0.703071 -0.14592 -0.21666 -0.035429  0.291569  1.0   
1    -0.277805  0.393668 -1.422331 -0.14592 -0.21666 -0.440378  0.291569  0.0   
2     0.966947  0.393668  0.703071 -0.14592 -0.21666 -0.035429  0.291569  0.0   
3    -0.900181  0.393668  0.703071 -0.14592 -0.21666  0.774468  0.291569  0.0   
4    -0.900181  0.393668  0.703071 -0.14592 -0.21666 -0.035429  0.291569  1.0   
...        ...       ...       ...      ...      ...       ...       ...  ...   
6508  2.211698  0.393668 -1.422331 -0.14592 -0.21666 -0.035429  0.291569  0.0   
6509  2.211698 -1.962621 -1.422331 -0.14592 -0.21666 -0.035429  0.291569  0.0   
6510 -0.900181  0.393668  0.703071 -0.14592 -0.21666  1.422386  0.291569  1.0   
6511  0.966947  0.393668  0.703071 -0.14592 -0.21666 -0.035429  0.291569  0.0   
6512  1.589322  0.393668 -1.422331 -0.14592 -0.21666 -0.035429  0.291569  0.0   

       15  
0     1.0  
1     0.0  
2     1.0  
3     0.0  
4     0.0  
...   ...  
6508  1.0  
6509  1.0  
6510  1.0  
6511  0.0  
6512  0.0  

[6513 rows x 16 columns]

### Naive Bayes

Verwendet Wahrscheinlichkeiten, um die Datenpunkte zu klassifizieren

Wenn ein Datenpunkt über 50% ist, wird er der einen Klasse zugewiesen, sonst der anderen

In [117]:
from sklearn.naive_bayes import GaussianNB

In [118]:
nb = GaussianNB()

In [119]:
nb_model = nb.fit(trainingL, trainingR)

In [121]:
predictions = nb_model.predict(testL)

In [122]:
evaluate(predictions)

Korrekte Vorhersagen: 83.1874712114233%
Falsche Vorhersagen: 16.812528788576692%


### Logistische Regression

Vergleicht eine Gesamtwahrscheinlichkeit mit einer Threshold (standardmäßig 50%)

In [123]:
from sklearn.linear_model import LogisticRegression

In [124]:
lr = LogisticRegression()

In [125]:
lr_model = lr.fit(trainingL, trainingR)

In [126]:
predictions = lr_model.predict(testL)

In [127]:
evaluate(predictions)

Korrekte Vorhersagen: 76.89236910793798%
Falsche Vorhersagen: 23.10763089206203%


### Support Vector Machines

Legt eine Hyperplane (H) zwischen alle Datenpunkte (mittig) und breitet im Anschluss ein Margin (M) aus

Alle Punkte die Links liegen werden zu Klasse 1 klassfiziert, alle anderen Punkte zur anderen Klasse

In [128]:
from sklearn.svm import SVC

In [129]:
svc = SVC()

In [130]:
svc_model = svc.fit(trainingL, trainingR)

In [131]:
predictions = svc_model.predict(testL)

In [132]:
evaluate(predictions)

Korrekte Vorhersagen: 78.78089973898356%
Falsche Vorhersagen: 21.21910026101643%


## Neurales Netzwerk

Besteht aus Knotenpunkten (Nodes), diese Nodes werden in Schichten angelegt und miteinander verbunden

Bei jedem Lernprozess werden die Daten durch dieses Netzwerk gesendet

Jede Node besteht aus einem Neuron

Neuron:
- Am Anfang werden die Inputs multipliziert und im Anschluss aufsummiert
- Danach wird diese Summe in eine Activation Function gesendet
- Activation Function: Normale Mathematische Funktion
- Am Ende des Neurons werden die Outputdaten an die nächste Schicht weitergegeben

Am Ende eines Durchgangs fängt der Lernprozess von vorne an

In [133]:
import tensorflow as tf

In [134]:
# keras: Tensorflow API, erleichtert das Erstellen von Modellen
model = tf.keras.Sequential([
    tf.keras.layers.Input(shape=(14,)),
    tf.keras.layers.Dense(16, activation="relu"),
    tf.keras.layers.Dense(16, activation="relu"),
    tf.keras.layers.Dense(1, activation="sigmoid")
])

In [135]:
model.compile(loss="binary_crossentropy")  # Wie gut ist die 1 und 0 Klassifizierung?

In [136]:
model.fit(trainingL,
          trainingR,
          verbose=1,  # Outputs des Lernprozesses anzeigen
          epochs=10)  # Anzahl Lernprozesse

Epoch 1/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - loss: 0.5332
Epoch 2/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.4028
Epoch 3/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3915
Epoch 4/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3897
Epoch 5/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3845
Epoch 6/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3845
Epoch 7/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3811
Epoch 8/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - loss: 0.3785
Epoch 9/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3762
Epoch 10/10
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3786


In [137]:
model.fit(trainingL,
          trainingR,
          verbose=1,  # Outputs des Lernprozesses anzeigen
          epochs=40)  # Anzahl Lernprozesse

Epoch 1/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3720
Epoch 2/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - loss: 0.3745
Epoch 3/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3706
Epoch 4/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3700
Epoch 5/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3747
Epoch 6/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3719
Epoch 7/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3686
Epoch 8/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3762
Epoch 9/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3746
Epoch 10/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3755
Epoch 11/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3718
Epoch 12/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - loss: 0.3771
Epoch 13/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3718
Epoch 14/40
1233/1233 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.3679
Epoch 15/40
1233/1233 ━━━━━━━

In [138]:
model.save("Income.keras")

In [140]:
predictions = model.predict(testL)

204/204 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


In [145]:
predInt = (predictions > 0.5).astype(int).reshape(-1)

In [149]:
pd.Series(predInt == testR.values).value_counts()

True     5218
False    1295
Name: count, dtype: int64

In [157]:
predInt

array([1, 0, 1, ..., 0, 0, 0])

In [158]:
counts = pd.Series(predInt == testR.values).value_counts()
print(f"Korrekte Vorhersagen: {counts[True] / len(testR) * 100}%")
print(f"Falsche Vorhersagen: {counts[False] / len(testR) * 100}%")

Korrekte Vorhersagen: 80.11668969752802%
Falsche Vorhersagen: 19.88331030247198%
